src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1481,6 +1481,21 @@ public void ApplyChangeToWorkspace(Action<Workspace> action)
             }
         }
 
+        /// <summary>
+        /// Applies a solution transformation to the workspace and triggers workspace changed event for specified <paramref name="projectId"/>.
+        /// The transformation shall only update the project of the solution with the specified <paramref name="projectId"/>.
+        /// </summary>
+        public void ApplyChangeToWorkspace(ProjectId projectId, Func<CodeAnalysis.Solution, CodeAnalysis.Solution> solutionTransformation)
+        {
+            lock (_gate)
+            {
+                if (SetCurrentSolutionWithLock(solutionTransformation, out var oldSolution, out var newSolution))
```

Comment:
is this dangerous/expensive to do while holding the lock? it's effectively an unbounded computation, no?

Reply:
Yes, it is dangerous. We may easily deadlock. See e.g. https://github.com/dotnet/roslyn/issues/42229

Reply:
Yes, the locking here is not ideal. But that's what we are doing all over this class. Here at least we control what the transformations are since we are in VS layer.